File: ruby/all-your-base/all_your_base.rb
class BaseConverter
  def self.convert(input_base, digits, output_base)
    # Binary is the smallest base we allow.
    return nil if input_base < 2 or output_base < 2
    # Parts can't be negative or greater than their base allows.
    return nil if digits.any? { |d| d < 0 || d >= input_base }
    return [] if digits.length == 0
    t = get_total(input_base, digits)
    to_base(output_base, t)
  end

  private

  def self.get_total(input_base, digits)
    digits
      .reverse
      .each_with_index
      .inject(0) { |acc, (i,j)| acc += i * input_base**j }
  end

  def self.to_base(output_base, total)
    out = []
    current = total
    while current >= output_base do
      out.push(current % output_base)
      current = current / output_base
    end
    out.push(current % output_base)
    out.reverse
  end
end

module BookKeeping
  VERSION = 1
end

File: javascript/all-your-base/all-your-base.js
const Converter = function() {};

Converter.prototype.convert = function(digits, inputBase, outputBase) {
  if (!Converter.isValidBase(inputBase)) {
    throw new Error('Wrong input base');
  }

  if (!Converter.isValidBase(outputBase)) {
    throw new Error('Wrong output base');
  }

  if (!Converter.isValidDigits(digits, inputBase)) {
    throw new Error('Input has wrong format');
  }

  return Converter.toBase(outputBase, Converter.getTotal(inputBase, digits));
};

Converter.isValidBase = function(base) {
  return base != undefined &&
    base > 1 &&
    Number.isInteger(base);
};

Converter.isValidDigits = function(digits, inputBase) {
  const isEmpty = digits.length == 0;
  const leadingZero = digits.length > 1 && digits[0] == 0;
  const hasNegatives = digits.some((i) => i < 0);
  const tooHighForBase = digits.some((i) => i >= inputBase);
  return !(isEmpty || leadingZero || hasNegatives || tooHighForBase);
};

Converter.getTotal = function(inputBase, digits) {
  return digits
    .reverse()
    .reduce(function(pV, cV, i) {
      return pV += cV * Math.pow(inputBase, i);
    }, 0);
};

Converter.toBase = function(outputBase, total) {
  let out = [];
  let current = total;
  while (current >= outputBase) {
    out.push(current % outputBase);
    current = Math.floor(current / outputBase);
  }
  out.push(current % outputBase);
  return out.reverse();
};

module.exports = Converter;
